src/core/lib/security/authorization/mock_cel_library/cel_value.h

```diff
@@ -0,0 +1,242 @@
+                            const CelList*, const CelMap*, const UnknownSet*,
+                            const CelError*>;
+
+ public:
+  // Metafunction providing positions corresponding to specific
+  // types. If type is not supported, compile-time error will occur.
+  template <class T>
+  using IndexOf = ValueHolder::IndexOf<T>;
+
+  // Enum for types supported.
+  enum class Type {
+    kBool = IndexOf<bool>::value,
+    kInt64 = IndexOf<int64_t>::value,
+    kUint64 = IndexOf<uint64_t>::value,
+    kDouble = IndexOf<double>::value,
+    kString = IndexOf<StringHolder>::value,
+    kBytes = IndexOf<BytesHolder>::value,
+    kMessage = IndexOf<const google::protobuf::Message*>::value,
+    kList = IndexOf<const CelList*>::value,
+    kMap = IndexOf<const CelMap*>::value,
+    kUnknownSet = IndexOf<const UnknownSet*>::value,
+    kError = IndexOf<const CelError*>::value,
+    kAny  // Special value. Used in function descriptors.
+  };
+
+  // Default constructor.
+  // Creates CelValue with null data type.
+  CelValue()
+      : CelValue(static_cast<const google::protobuf::Message*>(nullptr)) {}
+
+  // Returns Type that describes the type of value stored.
+  Type type() const { return Type(value_.index()); }
+
+  // We will use factory methods instead of public constructors
+  // The reason for this is the high risk of implicit type conversions
+  // between bool/int/pointer types.
+  // We rely on copy elision to avoid extra copying.
+  static CelValue CreateNull() {
+    return CelValue(static_cast<const google::protobuf::Message*>(nullptr));
+  }
+
+  static CelValue CreateBool(bool value) { return CelValue(value); }
+
+  static CelValue CreateInt64(int64_t value) { return CelValue(value); }
+
+  static CelValue CreateUint64(uint64_t value) { return CelValue(value); }
+
+  static CelValue CreateDouble(double value) { return CelValue(value); }
+
+  static CelValue CreateString(StringHolder holder) { return CelValue(holder); }
+
+  static CelValue CreateStringView(absl::string_view value) {
+    return CelValue(StringHolder(value));
+  }
+
+  static CelValue CreateString(const std::string* str) {
+    return CelValue(StringHolder(str));
+  }
+
+  static CelValue CreateBytes(BytesHolder holder) { return CelValue(holder); }
+
+  static CelValue CreateBytesView(absl::string_view value) {
+    return CelValue(BytesHolder(value));
+  }
+
+  static CelValue CreateBytes(const std::string* str) {
+    return CelValue(BytesHolder(str));
+  }
+
+  // CreateMessage creates CelValue from google::protobuf::Message.
+  // As some of CEL basic types are subclassing google::protobuf::Message,
+  // this method contains type checking and downcasts.
+  static CelValue CreateMessage(const google::protobuf::Message* value,
+                                google::protobuf::Arena* arena);
+
+  static CelValue CreateList(const CelList* value) {
+    CheckNullPointer(value, Type::kList);
```

Comment:
Remove these checks. 
Also, maybe we can just return CreateNull everywhere. WDYT?

Reply:
Thanks for the review. This approach sounds good. I have implemented it and also removed as much other unnecessary code as possible.